src/interp.rs

```diff
@@ -19,342 +19,101 @@
-        factor: usize,
-        /// Taps (prefer odd to increase zero coeffs)
-        taps: usize,
-        /// Number of channels
-        channels: u32,
-        /// Size of delay buffer
-        delay: usize,
-        /// List of subfilters (one for each factor)
-        filter: SmallVec<[Filter; 4]>,
-        /// List of delay buffers (one for each channel)
-        z: Box<[f32]>,
-        /// Current delay buffer index
-        zi: usize,
-    }
-
-    #[derive(Debug)]
-    struct Filter {
-        /// List of subfilter coefficients and corresponding delay indices
-        coeff: Box<[(f64, usize)]>,
-    }
-
-    impl Interp {
-        #[allow(clippy::many_single_char_names)]
-        pub fn new(taps: usize, factor: usize, channels: u32) -> Self {
-            let delay = (taps + factor - 1) / factor;
-
-            // Initialize the filter memory
-            // One subfilter per interpolation factor.
-            let mut filter = SmallVec::<[_; 4]>::new();
-
-            for _ in 0..factor {
-                filter.push(vec![]);
-            }
-
-            // One delay buffer per channel.
-            let z = vec![0.0; delay * channels as usize];
-
-            // Calculate the filter coefficients.
-            for j in 0..taps {
-                const ALMOST_ZERO: f64 = 0.000001;
-
-                // Calculate Hanning window
-                let w =
-                    0.5 * (1.0 - f64::cos(2.0 * f64::consts::PI * j as f64 / (taps - 1) as f64));
-
-                // Calculate sinc and apply hanning window
-                let m = j as f64 - (taps - 1) as f64 / 2.0;
-                let c = if m.abs() > ALMOST_ZERO {
-                    w * f64::sin(m * f64::consts::PI / factor as f64)
-                        / (m * f64::consts::PI / factor as f64)
-                } else {
-                    w
-                };
-
-                // Ignore any zero coeffs.
-                if c.abs() > ALMOST_ZERO {
-                    // Put the coefficient into the correct subfilter
-                    let f = j % factor;
-
-                    let f = &mut filter[f];
-                    f.push((c, j / factor));
-                }
-            }
-
-            Interp {
-                factor,
-                taps,
-                channels,
-                delay,
-                filter: filter
-                    .into_iter()
-                    .map(|f| Filter {
-                        coeff: f.into_boxed_slice(),
-                    })
-                    .collect(),
-                z: z.into_boxed_slice(),
-                zi: 0,
-            }
-        }
-    }
-
-    impl super::Interpolator for Interp {
-        fn get_factor(&self) -> usize {
-            self.factor
+use crate::utils::FrameAccumulator;
+use std::f64::consts::PI;
+
+const ALMOST_ZERO: f64 = 0.000001;
+const TAPS: usize = 48;
+
+macro_rules! interp_impl {
+    ( $name:ident, $factor:expr ) => {
+        #[derive(Debug, Clone)]
```

Comment:
Also here the question, why is `Clone` needed?

Reply:
Same answer; it is used by `vec!`.